From d95a8e1f4d42374508a326636a329e009d8bea6a Mon Sep 17 00:00:00 2001
From: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>
Date: Fri, 18 Nov 2022 05:20:19 -0800
Subject: [PATCH 3/4] ath12k: Program the RBM ID values in TCL RBM map register

Program the RBM ID values during TCL init and avoid providing
the rbm id values on a per packet basis

Signed-off-by: Balamurugan Mahalingam <quic_bmahalin@quicinc.com>

--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -508,6 +508,7 @@ static int ath12k_dp_srng_common_setup(s
 	struct hal_srng *srng;
 	int i, ret, tx_comp_ring_num;
 	u32 ring_hash_map;
+	u8 rbm_id;
 
 	ret = ath12k_dp_srng_setup(ab, &dp->wbm_desc_rel_ring,
 				   HAL_SW2WBM_RELEASE, 0, 0,
@@ -534,6 +535,7 @@ static int ath12k_dp_srng_common_setup(s
 
 	for (i = 0; i < DP_TCL_NUM_RING_MAX; i++) {
 		tx_comp_ring_num = ab->hal.ops->tcl_to_wbm_rbm_map[i].wbm_ring_num;
+		rbm_id = ab->hal.ops->tcl_to_wbm_rbm_map[i].rbm_id;
 
 		ret = ath12k_dp_srng_setup(ab, &dp->tx_ring[i].tcl_data_ring,
 					   HAL_TCL_DATA, i, 0,
@@ -552,6 +554,7 @@ static int ath12k_dp_srng_common_setup(s
 				    tx_comp_ring_num, ret);
 			goto err;
 		}
+		ath12k_hal_tx_config_rbm_mapping(ab, i, rbm_id, HAL_TCL_DATA);
 
 		srng = &ab->hal.srng_list[dp->tx_ring[i].tcl_data_ring.ring_id];
 	}
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -107,25 +107,6 @@ static void ath12k_hal_tx_cmd_ext_desc_s
 					   ti->encrypt_type);
 }
 
-static const struct ath12k_hal_tcl_to_wbm_rbm_map
-ath12k_hal_qcn92xx_tcl_to_wbm_rbm_map[DP_TCL_NUM_RING_MAX] = {
-	{
-		.wbm_ring_num = 0,
-		.rbm_id = HAL_RX_BUF_RBM_SW0_BM,
-	},
-	{
-		.wbm_ring_num = 1,
-		.rbm_id = HAL_RX_BUF_RBM_SW1_BM,
-	},
-	{
-		.wbm_ring_num = 2,
-		.rbm_id = HAL_RX_BUF_RBM_SW2_BM,
-	},
-	{
-		.wbm_ring_num = 4,
-		.rbm_id = HAL_RX_BUF_RBM_SW4_BM,
-	}
-};
 
 int ath12k_dp_tx_direct(struct ath12k_vif *arvif, struct sk_buff *skb)
 {
@@ -138,13 +119,11 @@ int ath12k_dp_tx_direct(struct ath12k_vi
 	struct dp_tx_ring *tx_ring;
 	void *hal_tcl_desc;
 	struct hal_tcl_data_cmd tcl_desc = { 0 };
-	u8 hal_ring_id, ring_id, rbm_id;
+	u8 hal_ring_id, ring_id;
 	int ret;
 
 	ring_id = smp_processor_id();
 
-	rbm_id = ath12k_hal_qcn92xx_tcl_to_wbm_rbm_map[ring_id].rbm_id;
-
 	tx_ring = &dp->tx_ring[ring_id];
 
 	tx_desc = ath12k_dp_tx_assign_buffer_nolock(dp, ring_id);
@@ -186,7 +165,6 @@ int ath12k_dp_tx_direct(struct ath12k_vi
 	tcl_desc.buf_addr_info.info1 = FIELD_PREP(BUFFER_ADDR_INFO1_ADDR,
 			((uint64_t)skb_cb->paddr >> HAL_ADDR_MSB_REG_SHIFT));
 	tcl_desc.buf_addr_info.info1 |=
-		FIELD_PREP(BUFFER_ADDR_INFO1_RET_BUF_MGR, rbm_id) |
 		FIELD_PREP(BUFFER_ADDR_INFO1_SW_COOKIE, tx_desc->desc_id);
 	tcl_desc.info0 = FIELD_PREP(HAL_TCL_DATA_CMD_INFO0_DESC_TYPE,
 				    HAL_TCL_DESC_TYPE_BUFFER) | arvif->desc.info0;
@@ -247,8 +225,6 @@ int ath12k_dp_tx(struct ath12k *ar, stru
 
 	ti.ring_id = smp_processor_id();
 
-	ti.rbm_id = ab->hal.ops->tcl_to_wbm_rbm_map[ti.ring_id].rbm_id;
-
 	tx_ring = &dp->tx_ring[ti.ring_id];
 
 	tx_desc = ath12k_dp_tx_assign_buffer_nolock(dp, ti.ring_id);
@@ -400,8 +376,6 @@ int ath12k_dp_tx(struct ath12k *ar, stru
 		FIELD_PREP(BUFFER_ADDR_INFO1_ADDR,
 			   ((uint64_t)ti.paddr >> HAL_ADDR_MSB_REG_SHIFT));
 	tcl_cmd->buf_addr_info.info1 |=
-		FIELD_PREP(BUFFER_ADDR_INFO1_RET_BUF_MGR,
-			   (ti.rbm_id)) |
 		FIELD_PREP(BUFFER_ADDR_INFO1_SW_COOKIE, ti.desc_id);
 
 	tcl_cmd->info0 = FIELD_PREP(HAL_TCL_DATA_CMD_INFO0_DESC_TYPE,
--- a/drivers/net/wireless/ath/ath12k/hal.h
+++ b/drivers/net/wireless/ath/ath12k/hal.h
@@ -112,6 +112,12 @@ struct ath12k_base;
 #define HAL_TCL1_RING_TP_OFFSET \
 		(HAL_TCL1_RING_TP - HAL_TCL1_RING_HP)
 
+#define HAL_TCL_RBM_MAPPING0_ADDR_OFFSET	0x00000088
+#define HAL_TCL_RBM_MAPPING_SHFT 4
+#define HAL_TCL_RBM_MAPPING_BMSK 0xF
+#define HAL_TCL_RBM_MAPPING_PPE2TCL_OFFSET  7
+#define HAL_TCL_RBM_MAPPING_TCL_CMD_CREDIT_OFFSET  6
+
 /* TCL STATUS ring address */
 #define HAL_TCL_STATUS_RING_BASE_LSB(ab) \
 	(ab)->hw_params.regs->hal_tcl_status_ring_base_lsb
--- a/drivers/net/wireless/ath/ath12k/hal_tx.c
+++ b/drivers/net/wireless/ath/ath12k/hal_tx.c
@@ -97,3 +97,26 @@ void ath12k_hal_tx_configure_bank_regist
 	ath12k_hif_write32(ab, HAL_TCL_SW_CONFIG_BANK_ADDR + 4 * bank_id,
 			   bank_config);
 }
+
+void ath12k_hal_tx_config_rbm_mapping(struct ath12k_base *ab,u8 ring_num,
+				      u8 rbm_id, int ring_type)
+{
+	u32 curr_map, new_map;
+
+	if (ring_type == HAL_PPE2TCL)
+		ring_num = ring_num + HAL_TCL_RBM_MAPPING_PPE2TCL_OFFSET;
+	else if (ring_type == HAL_TCL_CMD)
+		ring_num = ring_num + HAL_TCL_RBM_MAPPING_TCL_CMD_CREDIT_OFFSET;
+
+	curr_map = ath12k_hif_read32(ab, HAL_SEQ_WCSS_UMAC_TCL_REG +
+				     HAL_TCL_RBM_MAPPING0_ADDR_OFFSET);
+
+	/* Protect the other values and clear the specific fields to be updated */
+	curr_map &= (~(HAL_TCL_RBM_MAPPING_BMSK <<
+		      (HAL_TCL_RBM_MAPPING_SHFT * ring_num)));
+	new_map = curr_map | ((HAL_TCL_RBM_MAPPING_BMSK & rbm_id) <<
+			      (HAL_TCL_RBM_MAPPING_SHFT * ring_num));
+
+	ath12k_hif_write32(ab, HAL_SEQ_WCSS_UMAC_TCL_REG +
+			   HAL_TCL_RBM_MAPPING0_ADDR_OFFSET, new_map);
+}
--- a/drivers/net/wireless/ath/ath12k/hal_tx.h
+++ b/drivers/net/wireless/ath/ath12k/hal_tx.h
@@ -86,4 +86,6 @@ int ath12k_hal_reo_cmd_send(struct ath12
 			    struct ath12k_hal_reo_cmd *cmd);
 void ath12k_hal_tx_configure_bank_register(struct ath12k_base *ab, u32 bank_config,
 					   u8 bank_id);
+void ath12k_hal_tx_config_rbm_mapping(struct ath12k_base *ab,u8 ring_num,
+                                      u8 rbm_id, int ring_type);
 #endif
